tools/proofers/processtext.php

```diff
@@ -306,6 +306,82 @@ function leave_spellcheck_mode( $ppage )
     }
 }
 
+function attempt_to_save_as_done($ppage, $text_data, $and_do_another)
+// This is only an attempt, because a daily page limit might block the save,
+// or prevent further saves.
+// If there's a problem, this function does not return to the caller.
+{
+    global $code_url, $pguser, $projectid, $proj_state;
+
+    $round = get_Round_for_project_state($proj_state);
+
+    list($saved, $dpl_reached) = $ppage->attemptSaveAsDone($text_data, $pguser);
+
+    if (!$dpl_reached)
+    {
+        assert($saved);
+        return; // to let the caller do the appropriate normal thing
+    }
+
+    assert($dpl_reached);
+
+    if ($saved)
+    {
+        $title = _("Saved, but at limit");
+        $sentence = sprintf(
+            _("Your page has been saved as 'Done'. However, you have now reached the daily page limit for %s."),
+            $round->id
+        );
+    }
+    else
+    {
+        $ppage->saveAsInProgress($text_data, $pguser);
+        $title = _("Already at limit");
+        $sentence = sprintf(
+            _("Your page was saved as 'In Progress' rather than 'Done', because you have already reached the daily page limit for %s."),
+            $round->id
+        );
+    }
+
+    slim_header( $title );
+
+    echo "<p>$sentence</p>\n";
+    
+    echo "<p>"
+        . sprintf(
+            _("You will not be allowed to save any more pages in %s until the next server midnight."),
+            $round->id
+          )
+        . "</p>\n";
+
+    echo "<ul>\n"
+        .   "<li>"
+        .     sprintf(
+                _("Return to <a %s>the Project Page</a>"),
+                "href='$code_url/project.php?id=$projectid' target='_top'"
+              )
+        .   "</li>\n"
+        .   "<li>"
+        .     sprintf(
+                _("Return to <a %s>round %s</a>"),
+                "href='round.php?round_id={$round->id}' target='_top'",
+                $round->id
+              )
+        .   "</li>\n"
+        .   "<li>"
+        .     sprintf(
+                _("Return to <a %s>the Activity Hub</a>"),
+                "href='$code_url/activity_hub.php' target='_top'"
+              )
+        .   "</li>\n"
+        . "</ul>\n"
+        ;
+
+    slim_footer();
```

Comment:
No need to output footers - the shutdown function takes care of that for us.

Reply:
Done!